docs/src/description/constraints.md

```diff
@@ -2,8 +2,158 @@
+
+Transition constraints that are defined against auxiliary columns or that use random values from the built-in `$rand` array will be identified as auxiliary constraints.
+
+A transition constraint definition must:
+
+1. start with a block indentation and the `enf` keyword to indicate that the constraint must be _enforced_.
+2. continue with an equality expression that describes the constraint. The expression may include numbers, trace columns, periodic columns, random values, and any of the available [operations](./syntax.md#operations).
+3. end with a newline.
+
+### Current and next rows
+
+Transition constraints have access to values in the "current" row of the trace to which the constraint is being applied, as well as the "next" row of the trace. The value of a trace column in the next row is specified with the `'` postfix operator, as described by the [accessor syntax rules](./syntax.md#section-specific-accessors).
+
+### Simple example of transition constraints
+
+The following is a simple example of a valid `transition_constraints` block using values from the current and next rows of the main and auxiliary traces:
+
+```
+def TransitionConstraintsExample
+
+trace_columns:
+    main: [a, b]
+    aux: [p]
+
+public_inputs:
+    # at least one public inputs is required.
+    program_hash: [4]
+
+transition_constraints:
+    # these are main constraints. they both express the same constraint.
+    enf a' = a + 1
+    enf b' - b - 1 = 0
+
+    # this is an auxiliary constraint, since it uses an auxiliary trace column.
+    enf p = p' * a
+```
+
+### Periodic columns and random values
+
+Transition constraints can access the value of any periodic column in the current row, as well as random values provided by the verifier.
+
+To use periodic column values, the periodic column must be declared in the `periodic_columns` source section. The value in the current row can then be accessed by using the defined identifier of the periodic column.
+
+Random values can be accessed by using array indexing syntax on the `$rand` built-in, as described by the [accessor syntax rules](./syntax.md#section-specific-accessors).
+
+### Example of transition constraints with periodic columns and random values
+
+The following is an example of a valid `transition_constraints` block that uses periodic columns and random values:
+
+```
+def TransitionConstraintsExample
+
+trace_columns:
+    main: [a, b]
+    aux: [p0, p1]
+
+public_inputs:
+    # at least one public input section is required.
+    program_hash: [4]
+
+periodic_columns:
+    k: [1, 1, 1, 0]
+
+transition_constraints:
+    # this is a main constraint that uses a periodic column.
+    enf a' = k * a
+
+    # this is an auxiliary constraint that uses a periodic column.
+    enf p0' = k * p0
+
+    # these are auxiliary constraints that use random values from the verifier.
+    enf b = a + $rand[0]
```

Comment:
This probably should be invalid as this constraint should never be satisfied: the right hand side is an extension field element, while the left hand side is a base field element. If the value in `$rand[0]` is truly random, these will never be equal.

We'll need to think about a general rule which can describe this.

Reply:
Yeah - I aimed to demonstrate everything that's currently syntactically valid. This is a harder one to check than the boundary constraints, so it probably merits its own issue

Reply:
I've opened #47